community/security/src/main/java/org/neo4j/server/security/auth/FileRepositorySerializer.java

```diff
@@ -34,34 +38,59 @@
 
 public abstract class FileRepositorySerializer<S>
 {
-    public void saveRecordsToFile(Path recordsFile, Collection<S> records) throws IOException
+    private Random random = new SecureRandom();
+
+    public static void writeToFile( FileSystemAbstraction fs, File file, byte[] bytes ) throws IOException
     {
-        Path directory = recordsFile.getParent();
-        if ( !Files.exists( directory ) )
-        {
-            Files.createDirectories( directory );
-        }
+        OutputStream o = fs.openAsOutputStream( file, false );
+        o.write( bytes );
+        o.close();
+    }
+
+    public static byte[] readFromFile( FileSystemAbstraction fs, File file ) throws IOException
+    {
+        InputStream s = fs.openAsInputStream( file );
+        byte[] bytes = sun.misc.IOUtils.readFully( s, -1, false );
+        s.close();
+        return bytes;
+    }
+
+    public void saveRecordsToFile( FileSystemAbstraction fileSystem, File recordsFile, Collection<S> records ) throws
+            IOException
+    {
+        File tempFile = getTempFile( fileSystem, recordsFile );
 
-        Path tempFile = Files.createTempFile( directory, recordsFile.getFileName().toString() + "-", ".tmp" );
         try
         {
-            Files.write( tempFile, serialize( records ) );
-            Files.move( tempFile, recordsFile, ATOMIC_MOVE, REPLACE_EXISTING );
+            writeToFile( fileSystem, tempFile, serialize( records ) );
+            fileSystem.move( tempFile, recordsFile, ATOMIC_MOVE, REPLACE_EXISTING );
```

Comment:
We recover from transient errors during move, as tested in the FileUserRepository test. If neo4j is installed on a filesystem were an atomic move within the same directory is not possible then the flat-file based security is not recommended, and LDAP should be used instead.